### PR TITLE
SnapshotRenderingHelper: Fix interaction with performance priority mode

### PR DESCRIPTION
**Pull request by Popov72**

### CODE REVIEW

**Review by deltakosh — APPROVED** [review]

**Review by sebavan — APPROVED** [review]

File: packages/dev/core/src/Misc/snapshotRenderingHelper.ts [hunk 1]
```diff
@@ -134,6 +141,29 @@ export class SnapshotRenderingHelper {
...
+            // We reset the performance priority mode to that which it was before enabling snapshot rendering, but first set it to “BackwardCompatible” to allow the system to regenerate resources that may have been optimized for snapshot rendering.
+            // We'll then restore the original mode at the next frame.
+            this._scene.performancePriority = ScenePerformancePriority.BackwardCompatible;
+            if (this._currentPerformancePriorityMode !== ScenePerformancePriority.BackwardCompatible) {
+                this._pendingCurrentPerformancePriorityMode = this._currentPerformancePriorityMode;
+                this._scene.executeWhenReady(() => {
+                    this._executeAtFrame(
+                        this._engine.frameId + 2,
+                        () => {
+                            // if this._disableRenderingRefCount === 0, snapshot rendering has been enabled in the meantime, so do nothing
```
Comment (RaananW) [inline packages/dev/core/src/Misc/snapshotRenderingHelper.ts:155]:
```suggestion
                            // if this._disableRenderingRefCount === 0 snapshot rendering has been enabled in the meantime, so do nothing
```

forcing a new build (not really needed)

File: packages/dev/core/src/Misc/snapshotRenderingHelper.ts [hunk 2]
```diff
@@ -134,6 +141,29 @@ export class SnapshotRenderingHelper {
...
+            // We reset the performance priority mode to that which it was before enabling snapshot rendering, but first set it to “BackwardCompatible” to allow the system to regenerate resources that may have been optimized for snapshot rendering.
+            // We'll then restore the original mode at the next frame.
+            this._scene.performancePriority = ScenePerformancePriority.BackwardCompatible;
+            if (this._currentPerformancePriorityMode !== ScenePerformancePriority.BackwardCompatible) {
+                this._pendingCurrentPerformancePriorityMode = this._currentPerformancePriorityMode;
+                this._scene.executeWhenReady(() => {
+                    this._executeAtFrame(
+                        this._engine.frameId + 2,
+                        () => {
+                            // if this._disableRenderingRefCount === 0 snapshot rendering has been enabled in the meantime, so do nothing
```
Comment (RaananW) [inline packages/dev/core/src/Misc/snapshotRenderingHelper.ts:155]:
```suggestion
                            // if this._disableRenderingRefCount === 0, snapshot rendering has been enabled in the meantime, so do nothing
```

**Comment by RaananW** [comment]
Build manual - https://dev.azure.com/babylonjs/ContinousIntegration/_build/results?buildId=34582&view=results

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Misc/snapshotRenderingHelper.ts [hunk 3]
```diff
@@ -134,6 +141,29 @@ export class SnapshotRenderingHelper {
             return;
         }
 
+        if (this._disableRenderingRefCount === 0) {
+            // Snapshot rendering switches from enabled to disabled
+            // We reset the performance priority mode to that which it was before enabling snapshot rendering, but first set it to “BackwardCompatible” to allow the system to regenerate resources that may have been optimized for snapshot rendering.
+            // We'll then restore the original mode at the next frame.
+            this._scene.performancePriority = ScenePerformancePriority.BackwardCompatible;
+            if (this._currentPerformancePriorityMode !== ScenePerformancePriority.BackwardCompatible) {
+                this._pendingCurrentPerformancePriorityMode = this._currentPerformancePriorityMode;
```
Comment (ryantrem) [inline packages/dev/core/src/Misc/snapshotRenderingHelper.ts:150]: Seems like `pendingCurrentPerformancePriorityMode` can just be a local variable. It is set to undefined in the same place that disableREnderingRefCount is set to 0, and the code already checks the value of disableRenderingRefCount. In other words, pendingCurrentPerformancePriorityMode will be undefined iff disableRenderingRefCount is 0. Is that right, or did I miss something in the logic?

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15743/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15743/merge/testResults/webgl2playwright/index.html

File: packages/dev/core/src/Misc/snapshotRenderingHelper.ts (hunk 3 above)
Comment (Popov72) [reply packages/dev/core/src/Misc/snapshotRenderingHelper.ts:150]: I need this to be a class variable because I reuse it here in `enableSnapshotRendering()`:

```typescript
this._currentPerformancePriorityMode = this._pendingCurrentPerformancePriorityMode ?? this._scene.performancePriority;
```
I can't use only `this._scene.performancePriority` because the value will have been set to `Backward compatible` early in `disableSnapshotRendering()`, and if we don't reach the code that sets `this._scene.performancePriority` to the right value (because snapshot mode has been reenabled before we could reach this point - line 157), we need to set the right value to `this._currentPerformancePriorityMode` and not `Backward compatible`.

Comment (ryantrem) [reply packages/dev/core/src/Misc/snapshotRenderingHelper.ts:150]: Ah ok. I'm finding it a little hard to follow the logic, but I don't have ideas on how to simplify it.

**Review by ryantrem — APPROVED** [review]

**Comment by RaananW** [comment]
Approved here - https://dev.azure.com/babylonjs/ContinousIntegration/_build/results?buildId=34582&view=results
